File: flist.cpp
#include "flist.h"
#include <iostream>

void CreateFList(FList* flist) {
    flist -> size = 0;
    flist -> head = nullptr;
}

FNode* CreateFNode(string data) {
    FNode* newNode = new FNode;
    newNode -> data = data;
    newNode -> next = nullptr;
    return newNode;
}

void FPushHead(FList* flist, string data) {
    FNode* newNode = CreateFNode(data);
    newNode -> next = flist -> head;
    flist -> head = newNode;
    flist -> size++;
}

void FPushBack(FList* flist, string data) {
    FNode* newNode = CreateFNode(data);
    if (flist -> head == nullptr) {
        flist -> head = newNode;
    } else {
        FNode* current = flist -> head;
        while (current -> next) current = current -> next;
        current -> next = newNode;
    }
    flist -> size++;
}

void FPushBefore(FList* flist, string data, string element) {
    if (flist -> head -> data == element) {
        FPushHead(flist, data);
        return;
    }
    FNode* current = flist -> head;
    while (current -> next && current -> next -> data != element) {
        current = current -> next;
    }
    if (current -> next) {
        FNode* newNode = CreateFNode(data);
        newNode -> next = current -> next;
        current -> next = newNode;
        flist -> size++;
    }
}

void FPushAfter(FList* flist, string data, string element) {
    FNode* current = flist -> head;
    while (current) {
        if (current -> data == element) {
            FNode* newNode = CreateFNode(data);
            newNode -> next = current -> next;
            current -> next = newNode;
            flist -> size++;
            return;
        }
        current = current -> next;
    }
}

void FPopHead(FList* flist) {
    if (!flist -> head) return;
    FNode* temp = flist -> head;
    flist -> head = flist -> head -> next;
    delete temp;
    flist -> size--;
}

void FPopTail(FList* flist) {
    if (!flist -> head) return;
    if (!flist -> head -> next) {
        delete flist -> head;
        flist -> head = nullptr;
        flist -> size--;
        return;
    }
    FNode* current = flist -> head;
    while (current -> next -> next) current = current -> next;
    delete current -> next;
    current -> next = nullptr;
    flist -> size--;
}

void FPopElement(FList* flist, string element) {
    if (!flist -> head) return;
    if (flist -> head -> data == element) {
        FPopHead(flist);
        return;
    }
    FNode* current = flist -> head;
    while (current -> next && current -> next -> data != element) {
        current = current -> next;
    }
    if (current -> next) {
        FNode* temp = current -> next;
        current -> next = current -> next -> next;
        delete temp;
        flist -> size--;
    }
}

bool FSearch(FList* flist, string element) {
    FNode* current = flist -> head;
    while (current) {
        if (current -> data == element) return true;
        current = current -> next;
    }
    return false;
}

void PrintFList(FList* flist) {
    cout << "[";
    FNode* current = flist -> head;
    while (current) {
        cout << current -> data;
        if (current -> next) cout << ", ";
        current = current -> next;
    }
    cout << "]" << endl;
}

void DestroyFList(FList* flist) {
    FNode* current = flist -> head;
    while (current) {
        FNode* temp = current;
        current = current -> next;
        delete temp;
    }
    flist -> head = nullptr;
}

File: flist.go
package main

import "fmt"

type FNode struct {
	Data string
	Next *FNode
}

type FList struct {
	Size int
	Head *FNode
}

func CreateFList() *FList {
	return &FList{
		Size: 0,
		Head: nil,
	}
}

func CreateFNode(data string) *FNode {
	return &FNode{
		Data: data,
		Next: nil,
	}
}

func (flist *FList) PushHead(data string) {
	newNode := CreateFNode(data)
	newNode.Next = flist.Head
	flist.Head = newNode
	flist.Size++
}

func (flist *FList) PushBack(data string) {
	newNode := CreateFNode(data)
	if flist.Head == nil {
		flist.Head = newNode
	} else {
		current := flist.Head
		for current.Next != nil {
			current = current.Next
		}
		current.Next = newNode
	}
	flist.Size++
}

func (flist *FList) PushBefore(data, element string) {
	if flist.Head == nil {
		return
	}
	if flist.Head.Data == element {
		flist.PushHead(data)
		return
	}
	current := flist.Head
	for current.Next != nil && current.Next.Data != element {
		current = current.Next
	}
	if current.Next != nil {
		newNode := CreateFNode(data)
		newNode.Next = current.Next
		current.Next = newNode
		flist.Size++
	}
}

func (flist *FList) PushAfter(data, element string) {
	current := flist.Head
	for current != nil {
		if current.Data == element {
			newNode := CreateFNode(data)
			newNode.Next = current.Next
			current.Next = newNode
			flist.Size++
			return
		}
		current = current.Next
	}
}

func (flist *FList) PopHead() {
	if flist.Head == nil {
		return
	}
	flist.Head = flist.Head.Next
	flist.Size--
}

func (flist *FList) PopTail() {
	if flist.Head == nil {
		return
	}
	if flist.Head.Next == nil {
		flist.Head = nil
		flist.Size--
		return
	}
	current := flist.Head
	for current.Next.Next != nil {
		current = current.Next
	}
	current.Next = nil
	flist.Size--
}

func (flist *FList) PopElement(element string) {
	if flist.Head == nil {
		return
	}
	if flist.Head.Data == element {
		flist.PopHead()
		return
	}
	current := flist.Head
	for current.Next != nil && current.Next.Data != element {
		current = current.Next
	}
	if current.Next != nil {
		current.Next = current.Next.Next
		flist.Size--
	}
}

func (flist *FList) Search(element string) bool {
	current := flist.Head
	for current != nil {
		if current.Data == element {
			return true
		}
		current = current.Next
	}
	return false
}

func (flist *FList) Print() {
	fmt.Print("[")
	current := flist.Head
	for current != nil {
		fmt.Print(current.Data)
		if current.Next != nil {
			fmt.Print(", ")
		}
		current = current.Next
	}
	fmt.Println("]")
}